docker/docker_client.go

```diff
@@ -407,7 +490,8 @@ func (c *dockerClient) detectProperties(ctx context.Context) error {
 		if !isV1 && c.ctx != nil && c.ctx.DockerInsecureSkipTLSVerify {
 			isV1 = pingV1("http")
 		}
-		if isV1 {
+		// can talk to v1 registry if doing a search
+		if isV1 && !isSearch {
```

Comment:
Maybe I’m being stupid but I can’t see how this works; here `err != nil` necessarily because we got into this branch, so this only changes what is the reported error but not that the method fails, i.e. `newLoginClient.makeRequest` should AFAICT fail for V1.

Reply:
… Ah, I see what’s going on: The currently relevant registries like `docker.io` and `registry.access.redhat.com` also/primarily serve content using the docker/distribution `/v2/` API as well, so we _never_ get here.

If we are happy enough to limit ourselves such V2 + `/v1/search` registries, the `allowV1` boolean does not need to exist at all.

---

BTW the https://docs.docker.com/v1.4/reference/api/docker-io_api/ document you found says that only `Authorization: Basic` is supported, making both the ping and maybe all of `setupRequestAuth` seemingly unnecessary. OTOH https://github.com/docker/docker-registry/blob/master/docker_registry/toolkit.pyhttps://github.com/docker/docker-registry/blob/master/docker_registry/toolkit.py#L276L276 does something entirely different, an `X-Signature` header or `Authorization: Token` – and `/v1/search` does not require any authentication anyway.

Similarly `/v1/search` requires authentication on neither `index.docker.io` nor `r.a.rh.com`.

moby/moby seems to support _both_ v2-like and v1-like authentication for the `/v1/search` URL. I guess we’ll eventually find out which ones, if any, matter; that doesn't need to block this PR.